src/coreclr/jit/lowerxarch.cpp

```diff
@@ -6313,38 +6313,40 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
                 {
                     if ((intrinsicId >= NI_FMA_MultiplyAdd) && (intrinsicId <= NI_FMA_MultiplySubtractNegatedScalar))
                     {
-                        bool supportsRegOptional = false;
+                        bool     supportsRegOptional  = false;
+                        bool     supportsRegOptional2 = false;
+                        unsigned overwrittenOpNum     = 0;
+                        LIR::Use use;
 
-                        if (IsContainableHWIntrinsicOp(node, op3, &supportsRegOptional))
+                        if (BlockRange().TryGetUse(node, &use))
                         {
-                            // 213 form: op1 = (op2 * op1) + [op3]
-                            MakeSrcContained(node, op3);
+                            overwrittenOpNum = node->GetOverwrittenOpNumForFMA(use.User(), op1, op2, op3);
                         }
-                        else if (IsContainableHWIntrinsicOp(node, op2, &supportsRegOptional))
+
+                        if (overwrittenOpNum != 1 && IsContainableHWIntrinsicOp(node, op1, &supportsRegOptional) &&
```

Comment:
I see what you are doing here - If a operand is containable and if it is not the overwritten operand, then we mark it as contained, but the logic will be clear if you flip the conditions inside `if()` i.e. do something like:

```c++
// This will say that if op2 is containable and it is not overwritten, then make it contained. Easy readiblity.
if (IsContainableHWIntrinsicOp(node, op2, &supportsRegOptional2) && overwrittenOpNum != 2)
{ .. }
```

I also didn't understand the last `else if` and `else` for marking `op2` vs. `op3` as regoptional. Can we separate it out somehow to make it more clear? 

Also couple of comments will be helpful here.

Reply:
When I saw `GetOverwrittenOpNumForFMA()` I thought that there will be checking with regards to `0`...Can you highlight clearly in one of the branch or add a comment as what happens when none of the operands are overwritten and none of them are lastUse (i.e. the case when `GetOverwrittenOpNumForFMA()` returns `0`)?